CONTRIBUTING_A_NEW_QUEST.md

```diff
@@ -0,0 +1,507 @@
+
+You can look at what was changed to achieve the goal, where relevant code is located. How much coding was needed to implement something? And what kind of comments are typical, how long one needs to wait for maintainers and so on.
+
+This can be also used to locate relevant code, especially helpful if some change needs to be done in multiple files.
+
+# Copying
+
+Copy the relevant quest folder. Some contain multiple quests, in such case copy only what you need.
+
+Some quests are entirely defined in a single file, some have additional answer class, custom interface or utility classes.
+
+# Adjust the copy
+
+After the quest is copied it is necessary to adjust it a bit.
+
+Change its class name and the file name to the new one.
+
+In copied code change package info (things like `package de.westnordost.streetcomplete.quests.defibrillator` at the top) to match the new folder containing the quest.
+
+# Add the quest to the list of active ones
+
+Adjust [QuestsModule.kt](app/src/main/java/de/westnordost/streetcomplete/quests/QuestsModule.kt) file. It contains a big list of active quests, ordered by priority. Read [what governs their priority](https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/QuestsModule.kt#L169-L192) but do not worry too much, it can be tweaked later.
+
+Add your quest to the list so that it will be loaded by the app.
+
+## Quest anatomy
+See for example [simple yes/no quest asking whether AED is indoor or outdoor](app/src/main/java/de/westnordost/streetcomplete/quests/defibrillator/AddIsDefibrillatorIndoor.kt).
+
+
+TODO: maybe linking [app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/OsmElementQuestType.kt](app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/OsmElementQuestType.kt) would be better? And extend documentation there?
+
+### Element selection
+elementFilter property defines nodes, ways and relations which will be selected for a given quest. It is an element selection used by OsmFilterQuestType.
+
+```
+"""
+        nodes with
+         emergency = defibrillator
+         and access !~ private|no
+         and !indoor
+"""
+``` 
+
+This query will be limited to nodes (`nodes with`), which fulfill some requirements.
+
+- `emergency = defibrillator` tag must be present
+- `access` tag must not have values `private` or `no` to skip ones where mapper will be unable to survey (`!~ private|no` - to be more specific `optionA|optionB` is treated like `^optionA|optionB$` [regexp](https://en.wikipedia.org/wiki/Regular_expression))
+- `indoor` key must not be present at all, to show only ones where this tag is still missing
+
+It is specified as a string, in syntax specific to StreetComplete. You can look around some quests to see how it works. If you are trying to implement a new quest and you got stuck here, [open a new issue](https://github.com/streetcomplete/StreetComplete/issues) to request more thorough documentation here.
+
```

Comment:
It should be emphasized that in `elementFilter` syntax (unlike vast majority of other programming languages), there is **no precendence of `and` over `or`**,  so users should use a lot of parenthesis, or it will lead to puzzling bugs like https://github.com/streetcomplete/StreetComplete/pull/3784#issuecomment-1053674700

Reply:
You mean in the parsing code? Sure, either of those would be ideal fix by me! But if that is too complicated to implement, it should at least be documented here in the docs.

Reply:
> You mean in the parsing code?

Yes, I think that SC crashing with error on start would be preferable and avoid troubles you had

Reply:
I opened https://github.com/streetcomplete/StreetComplete/issues/3838

Reply:
This is wrong. There is operator precedence.